File: backend/app/models/metrics.py
from datetime import datetime
from enum import Enum
from typing import Any

from pydantic import BaseModel, Field


class SystemStatus(str, Enum):
    HEALTHY = "healthy"
    WARNING = "warning"
    CRITICAL = "critical"
    OFFLINE = "offline"


class SystemType(str, Enum):
    LINUX = "linux"
    DOCKER = "docker"
    QBITTORRENT = "qbittorrent"
    UNIFI = "unifi"
    UNAS = "unas"


class CpuMetrics(BaseModel):
    percent: float = 0.0
    cores: list[float] = Field(default_factory=list)
    load_avg: list[float] = Field(default_factory=list)


class MemoryMetrics(BaseModel):
    total: int = 0
    used: int = 0
    available: int = 0
    percent: float = 0.0


class DiskMetrics(BaseModel):
    mount: str
    total: int = 0
    used: int = 0
    free: int = 0
    percent: float = 0.0


class NetworkMetrics(BaseModel):
    bytes_sent: int = 0
    bytes_recv: int = 0
    upload_rate: float = 0.0
    download_rate: float = 0.0


class TemperatureMetrics(BaseModel):
    label: str
    current: float
    high: float | None = None
    critical: float | None = None


class LinuxMetrics(BaseModel):
    cpu: CpuMetrics = Field(default_factory=CpuMetrics)
    memory: MemoryMetrics = Field(default_factory=MemoryMetrics)
    disks: list[DiskMetrics] = Field(default_factory=list)
    network: NetworkMetrics = Field(default_factory=NetworkMetrics)
    temperatures: list[TemperatureMetrics] = Field(default_factory=list)
    uptime: int = 0


class ContainerMetrics(BaseModel):
    id: str
    name: str
    image: str
    status: str
    state: str
    cpu_percent: float = 0.0
    memory_usage: int = 0
    memory_limit: int = 0
    memory_percent: float = 0.0


class DockerMetrics(BaseModel):
    containers: list[ContainerMetrics] = Field(default_factory=list)
    running_count: int = 0
    stopped_count: int = 0
    total_count: int = 0


class TorrentInfo(BaseModel):
    name: str
    size: int
    progress: float
    download_speed: int
    upload_speed: int
    state: str
    eta: int | None = None


class QBittorrentMetrics(BaseModel):
    download_speed: int = 0
    upload_speed: int = 0
    active_downloads: int = 0
    active_uploads: int = 0
    total_torrents: int = 0
    torrents: list[TorrentInfo] = Field(default_factory=list)


class UnifiDevice(BaseModel):
    name: str
    mac: str
    model: str
    status: str
    uptime: int = 0


class UnifiMetrics(BaseModel):
    devices: list[UnifiDevice] = Field(default_factory=list)
    clients_count: int = 0
    guests_count: int = 0
    wan_download: float = 0.0
    wan_upload: float = 0.0


class StoragePool(BaseModel):
    name: str
    status: str
    total: int = 0
    used: int = 0
    available: int = 0
    percent: float = 0.0


class StorageDisk(BaseModel):
    name: str
    model: str
    serial: str
    status: str
    temperature: float | None = None


class UNASMetrics(BaseModel):
    pools: list[StoragePool] = Field(default_factory=list)
    disks: list[StorageDisk] = Field(default_factory=list)
    total_capacity: int = 0
    used_capacity: int = 0


class SystemMetrics(BaseModel):
    id: str
    name: str
    type: SystemType
    status: SystemStatus = SystemStatus.OFFLINE
    last_updated: datetime = Field(default_factory=datetime.utcnow)
    error: str | None = None
    metrics: (
        LinuxMetrics
        | DockerMetrics
        | QBittorrentMetrics
        | UnifiMetrics
        | UNASMetrics
        | dict[str, Any]
    ) = Field(default_factory=dict)


class MetricsUpdate(BaseModel):
    type: str = "metrics_update"
    timestamp: datetime = Field(default_factory=datetime.utcnow)
    systems: dict[str, SystemMetrics] = Field(default_factory=dict)

File: backend/src/models/metrics.rs
use serde::{Deserialize, Serialize};

// --- Linux ---

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct CpuMetrics {
    pub percent: f64,
    #[serde(default)]
    pub cores: Vec<f64>,
    #[serde(default)]
    pub load_avg: Vec<f64>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct MemoryMetrics {
    pub total: i64,
    pub used: i64,
    pub available: i64,
    pub percent: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DiskMetrics {
    pub mount: String,
    #[serde(default)]
    pub total: i64,
    #[serde(default)]
    pub used: i64,
    #[serde(default)]
    pub free: i64,
    #[serde(default)]
    pub percent: f64,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct NetworkMetrics {
    pub bytes_sent: i64,
    pub bytes_recv: i64,
    pub upload_rate: f64,
    pub download_rate: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TemperatureMetrics {
    pub label: String,
    pub current: f64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub high: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub critical: Option<f64>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct LinuxMetrics {
    pub cpu: CpuMetrics,
    pub memory: MemoryMetrics,
    #[serde(default)]
    pub disks: Vec<DiskMetrics>,
    pub network: NetworkMetrics,
    #[serde(default)]
    pub temperatures: Vec<TemperatureMetrics>,
    #[serde(default)]
    pub uptime: i64,
}

// --- Docker ---

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ContainerMetrics {
    pub id: String,
    pub name: String,
    pub image: String,
    pub status: String,
    pub state: String,
    #[serde(default)]
    pub cpu_percent: f64,
    #[serde(default)]
    pub memory_usage: i64,
    #[serde(default)]
    pub memory_limit: i64,
    #[serde(default)]
    pub memory_percent: f64,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct DockerMetrics {
    #[serde(default)]
    pub containers: Vec<ContainerMetrics>,
    #[serde(default)]
    pub running_count: i32,
    #[serde(default)]
    pub stopped_count: i32,
    #[serde(default)]
    pub total_count: i32,
}

// --- qBittorrent ---

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TorrentInfo {
    pub name: String,
    pub size: i64,
    pub progress: f64,
    pub download_speed: i64,
    pub upload_speed: i64,
    pub state: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub eta: Option<i64>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct QBittorrentMetrics {
    #[serde(default)]
    pub download_speed: i64,
    #[serde(default)]
    pub upload_speed: i64,
    #[serde(default)]
    pub active_downloads: i32,
    #[serde(default)]
    pub active_uploads: i32,
    #[serde(default)]
    pub total_torrents: i32,
    #[serde(default)]
    pub torrents: Vec<TorrentInfo>,
}

// --- UniFi ---

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UnifiDevice {
    pub name: String,
    pub mac: String,
    pub model: String,
    pub status: String,
    #[serde(default)]
    pub uptime: i64,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct UnifiMetrics {
    #[serde(default)]
    pub devices: Vec<UnifiDevice>,
    #[serde(default)]
    pub clients_count: i32,
    #[serde(default)]
    pub guests_count: i32,
    #[serde(default)]
    pub wan_download: f64,
    #[serde(default)]
    pub wan_upload: f64,
}

// --- UNAS ---

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StoragePool {
    pub name: String,
    pub status: String,
    #[serde(default)]
    pub total: i64,
    #[serde(default)]
    pub used: i64,
    #[serde(default)]
    pub available: i64,
    #[serde(default)]
    pub percent: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StorageDisk {
    pub name: String,
    pub model: String,
    pub serial: String,
    pub status: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub temperature: Option<f64>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct UNASMetrics {
    #[serde(default)]
    pub pools: Vec<StoragePool>,
    #[serde(default)]
    pub disks: Vec<StorageDisk>,
    #[serde(default)]
    pub total_capacity: i64,
    #[serde(default)]
    pub used_capacity: i64,
}
